From 12fe5695802d3852c7e67fffcc74f8b3c46698cc Mon Sep 17 00:00:00 2001
From: Tushar Darote <quic_tdarote@quicinc.com>
Date: Thu, 29 Jan 2026 18:25:22 +0530
Subject: [PATCH 5/9] c: Force delegate symbols from shared library

Add linker options to ensure delegate create/delete functions are resolved
from shared libraries rather than statically linked. Handles XNNPACK, external
delegate, and GPU delegates with platform-specific flags for Windows, macOS,
and Linux.

Upstream-Status: Submitted [https://github.com/tensorflow/tensorflow/pull/109379]
Signed-off-by: Tushar Darote <quic_tdarote@quicinc.com>
---
 tensorflow/lite/c/CMakeLists.txt | 42 ++++++++++++++++++++++++++++++++
 1 file changed, 42 insertions(+)

diff --git a/tensorflow/lite/c/CMakeLists.txt b/tensorflow/lite/c/CMakeLists.txt
index 931f4372c5f..9fcafb14495 100644
--- a/tensorflow/lite/c/CMakeLists.txt
+++ b/tensorflow/lite/c/CMakeLists.txt
@@ -80,10 +80,52 @@ if (TFLITE_C_BUILD_SHARED_LIBS)
   if (WIN32)
     target_compile_definitions(tensorflowlite_c PRIVATE TFL_COMPILE_LIBRARY)
     target_compile_definitions(tensorflow-lite PRIVATE TFL_COMPILE_LIBRARY)
+    # Force the linker to treat delegate create and delete as undefined
+    #   to pull them from static to shared library
+    if (TFLITE_ENABLE_XNNPACK)
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteXNNPackDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteXNNPackDelegateDelete")
+    endif()
+    if (TFLITE_ENABLE_EXTERNAL_DELEGATE)
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteExternalDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteExternalDelegateDelete")
+    endif()
+    if (CMAKE_SYSTEM_NAME MATCHES "Android" OR TFLITE_ENABLE_GPU)
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteGpuDelegateV2Create")
+      target_link_options(tensorflowlite_c PRIVATE "/INCLUDE:TfLiteGpuDelegateV2Delete")
+    endif()
   elseif (APPLE)
     target_link_options(tensorflowlite_c PRIVATE "-Wl,-exported_symbols_list,${TFLITE_SOURCE_DIR}/c/exported_symbols.lds")
+    # Force the linker to treat delegate create and delete as undefined
+    #   to pull them from static to shared library
+    if (TFLITE_ENABLE_XNNPACK)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteXNNPackDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteXNNPackDelegateDelete")
+    endif()
+    if (TFLITE_ENABLE_EXTERNAL_DELEGATE)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteExternalDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteExternalDelegateDelete")
+    endif()
+    if (CMAKE_SYSTEM_NAME MATCHES "Android" OR TFLITE_ENABLE_GPU)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteGpuDelegateV2Create")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,_TfLiteGpuDelegateV2Delete")
+    endif()
   else ()
     target_link_options(tensorflowlite_c PRIVATE "-Wl,--version-script,${TFLITE_SOURCE_DIR}/c/version_script.lds")
+    # Force the linker to treat delegate create and delete as undefined
+    #   to pull them from static to shared library
+    if (TFLITE_ENABLE_XNNPACK)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteXNNPackDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteXNNPackDelegateDelete")
+    endif()
+    if (TFLITE_ENABLE_EXTERNAL_DELEGATE)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteExternalDelegateCreate")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteExternalDelegateDelete")
+    endif()
+    if (CMAKE_SYSTEM_NAME MATCHES "Android" OR TFLITE_ENABLE_GPU)
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteGpuDelegateV2Create")
+      target_link_options(tensorflowlite_c PRIVATE "-Wl,-u,TfLiteGpuDelegateV2Delete")
+    endif()
   endif()
 endif()
 
-- 
2.34.1

